### PR TITLE
fix(cfb): recover interceptor credit + add play-shape drift check

### PR DESCRIPTION
**Pull request by saiemgilani**
## Summary

Two credit-assignment defects surfaced by a full-history census of play-type x flag combinations and participant population rates (3,145,840 plays, 2004-2025), plus a new check for the class of problem a written rule cannot catch.

### 1. Interceptor credit was broken two ways

`interception_player_name` was **99.2% null on the 29,216 Interception Return plays** despite play text that plainly names the interceptor — and the 0.8% that *did* populate held corrupted spans (`"by Hosea Wheeler at the WKU 47"` instead of `"Hosea Wheeler"`).

Root cause: the pattern required a `by` clause that newer ESPN texts drop, and its `intercepted (.+)` capture swallowed the whole narrative tail. The replacement handles both era shapes (`intercepted by <Name>` and `pass intercepted <Name> return ...`), optional `#NN` jersey prefixes, and bare initials (`S Castille`), terminating on the lowercase tail via the case-toggle idiom (Rust regex has no lookaround). **Measured: 99.9% of interception texts that name anyone at all now extract** — the residue (4,628 rows) are texts that never name the interceptor.

### 2. Those nameless texts are now covered by participants — and a pick is no longer a PBU

ESPN ships no `interceptor` participant type; on an interception it files the defender as `pass_defender`. The old mapping sent that to `pass_breakup_player_name` on **every** play, so picks were credited as pass breakups and the interceptor stayed null.

`__join_participants` now routes `pass_defender` to interception credit on interception plays and keeps PBU attribution elsewhere. **Verified against real data**: across the 313 Interception-Return-Touchdown rows carrying both a text-extracted interceptor and a `pass_defender` participant, the two agree on last name **93.9%** of the time; the residue is suffix/nickname variance of the same athlete (`Desmond King` vs `Desmond King II`), not a different player.

### 3. New `combo_drift` check

The definitional rules assert relationships we already know. This asks the complementary question: **has the data produced a play shape we have never seen?** A committed snapshot enumerates all **504 (play type x 25 mechanics flags) combinations** observed across 2004-2025; anything outside it is a WARN with decoded samples — a real new football situation, an upstream vocabulary change, or a parser regression, none of which a pre-written rule catches. Refresh is explicit (`--refresh`), never automatic.

Also extends the `cfb_pbp` projected read with the mechanics flags the signature needs.

## Testing
- 793 passed across `tests/cfb/` + `tests/contracts/` (17 new: 5 interceptor extraction, 5 participant routing, 7 combo drift)
- Interceptor extraction and the participant-routing agreement rate both measured against the full published history, not fixtures

## Summary by Sourcery

Fix interception credit assignment in college football play-by-play and introduce a validation check for novel play-type/flag combinations.

Bug Fixes:
- Correct interceptor name extraction from play-by-play text across historical interception return plays, including modern formats without a 'by' clause and various name/jersey shapes.
- Route ESPN `pass_defender` participants to interception credit on interception plays while restricting pass breakup credit to non-interception plays so picks are no longer treated as pass breakups.
- Ensure participant-based interceptor data augments rather than overwrites valid text-extracted interceptor names.

Enhancements:
- Add a combo drift validation check that detects previously unseen play-type and mechanics-flag combinations against a committed historical snapshot, with tooling to build and refresh that snapshot.
- Expose mechanics flag columns in the CFB play-by-play validation registry needed to form the combo drift signature alphabet.

Tests:
- Expand CFB player column tests to cover interceptor text extraction edge cases and null behavior when the text does not name an interceptor.
- Add focused tests for interceptor-credit routing through `__join_participants`, including interception vs non-interception plays and mixed frames.
- Introduce contract tests for the combo drift validation check, its snapshot handling, and the shipped CFB play-by-play combination snapshot.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved interception-player recognition across older and newer play descriptions, including jersey-number and abbreviated-name formats.
  * Interception plays now correctly credit the defender with the interception rather than a pass breakup.
  * Preserved pass-breakup attribution for non-interception plays and retained extracted names when participant data is unavailable.

* **Validation**
  * Added combo-drift checks to identify previously unseen play-type and mechanics combinations.
  * Expanded play-by-play consistency checks and metadata coverage for scoring, turnovers, kicks, punts, and possession changes.

### CODE REVIEW

File: sportsdataverse/cfb/cfb_pbp.py [hunk 1]
```diff
@@ -6801,6 +6817,41 @@ def __join_participants(self, play_df):
...
+                    coalesce_exprs.append(
+                        pl.when(is_int)
+                        .then(pl.col("pass_breakup_player_name"))
+                        .otherwise(
+                            pl.coalesce(
+                                pl.col(pd_part).str.strip_chars(),
+                                pl.col("pass_breakup_player_name"),
+                            ),
+                        )
+                        .alias("pass_breakup_player_name"),
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply sportsdataverse/cfb/cfb_pbp.py:6858]: Declined — this would delete real credit data. A pass can be tipped by one defender and intercepted by another (tip drill), and ESPN narrates exactly that: measured across 2004-2025, **78 interception plays carry "broken up by" in the text, and all 78 also say "intercepted", naming two distinct defenders** — e.g. `pass intercepted by #25 K.Bretz at SDSU01 broken up by #2 M.Taylor`. Those breakups are genuine.

The bug this PR fixes was the *participant* name being written to PBU on every play; on an interception that name IS the interceptor, so it is now routed to interception credit and never injected into PBU. What remains on an int play is only a regex-extracted PBU from the play text, which is correct. Semantics documented in 98607716.

File: tools/validation/checks/combo_drift.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                locator={"missing_columns": missing},
+                needs_judgment=True,
+            )
+        ]
+
+    observed = frame.with_columns(_signature_expr(type_col, snap_flags)).group_by("_combo").agg(pl.len().alias("n"))
+    known = set(snap.get("combos", {}))
+    new = observed.filter(~pl.col("_combo").is_in(list(known))).sort("n", descending=True)
+    if new.height == 0:
+        return []
```
[the remark this answers is not in the document]
Comment (saiemgilani) [reply tools/validation/checks/combo_drift.py:176]: Fixed in 98607716. `_combo` itself is never null (the type column is `fill_null("<null>")` and each flag yields a "0"/"1" literal), so the null-propagation path does not arise — but an *empty* snapshot would have made `is_in([])` return False for every row and flagged the entire frame as drift. That is now caught explicitly: an empty snapshot is treated as corrupt and reported with a regenerate instruction rather than compared against. Regression test added.
